File: python/exclusivequeue.py
from threading import *

leaderQueue = Semaphore(0)
followerQueue = Semaphore(0)
mutex = Semaphore(1)
rendezvous = Semaphore(0)
leaders = followers = 0
threads = []
count = 0
numthreads = 10

def leader():
    mutex.acquire()
    global count, followers, leaders

    if followers > 0:
        followers -= 1
        followerQueue.release()
    else:
        leaders += 1
        mutex.release()
        leaderQueue.acquire()

    count += 1
    print(f"count: {count}\n")
    rendezvous.acquire()
    mutex.release()
    

def follower():
    mutex.acquire()
    global count, followers, leaders

    if leaders > 0:
        leaders -= 1
        leaderQueue.release()
    else:
        followers += 1
        mutex.release()
        followerQueue.acquire()
    
    count -= 1
    print(f"count: {count}\n")
    rendezvous.release()

for i in range(numthreads):
    threads.append(Thread(target=leader))

for i in range(numthreads):
    threads.append(Thread(target=follower))

for thread in threads:
    thread.start()
    
for thread in threads:
    thread.join()

print("All threads have finished.")

File: C++/exclusivequeue.cpp
#include <stdio.h>
#include <semaphore.h>
#include <thread>
#include <vector>

using std::vector, std::thread;

sem_t leaderQueue;
sem_t followerQueue;
sem_t mutex;
sem_t rendezvous;
int counter = 0;
int leaders = 0;
int followers = 0;
int numthreads = 10;
vector<thread> threads;

void leader() {
    sem_wait(&mutex);

    if (followers > 0) {
        followers--;
        sem_post(&followerQueue);
    }
    else {
        leaders++;
        sem_post(&mutex);
        sem_wait(&leaderQueue);
    }

    counter++;
    printf("count: %d\n", counter);
    sem_wait(&rendezvous);
    sem_post(&mutex);
}

void follower() {
    sem_wait(&mutex);

    if (leaders > 0) {
        leaders--;
        sem_post(&leaderQueue);
    }
    else {
        followers++;
        sem_post(&mutex);
        sem_wait(&followerQueue);
    }

    counter--;
    printf("count: %d\n", counter);
    sem_post(&rendezvous);
}

int main() {
    sem_init(&leaderQueue, 0, 0);
    sem_init(&followerQueue, 0, 0);
    sem_init(&mutex, 0, 1);
    sem_init(&rendezvous, 0, 0);

    for (int i = 0; i < numthreads; ++i) {
        threads.push_back(thread(leader));
    }

    for (int i = 0; i < numthreads; ++i) {
        threads.push_back(thread(follower));
    }

    for (thread &t : threads) {
        t.join();
    }

    sem_destroy(&leaderQueue);
    sem_destroy(&followerQueue);
    sem_destroy(&mutex);
    sem_destroy(&rendezvous);

    return 0;
}